File: kt/src/main/kotlin/com/github/ephemient/aoc2025/Day4.kt
package com.github.ephemient.aoc2025

fun day4(input: String): Pair<Int, Int> {
    var points = buildSet {
        for ((y, line) in input.lines().withIndex()) {
            for ((x, char) in line.withIndex()) {
                if (char == '@') add(x to y)
            }
        }
    }
    val initial = points.size
    points = step(points) ?: return 0 to 0
    val part1 = initial - points.size
    while (true) points = step(points) ?: break
    return part1 to initial - points.size
}

private fun step(points: Set<Pair<Int, Int>>): Set<Pair<Int, Int>>? = buildSet {
    var found = false
    points.filterTo(this) { (x, y) ->
        val count = (x - 1..x + 1).sumOf { x -> (y - 1..y + 1).count { y -> x to y in points } }
        if (count > 4) {
            true
        } else {
            found = true
            false
        }
    }
    if (!found) return null
}

File: rs/src/day4.rs
use std::collections::HashSet;

pub fn solve(input: &str) -> (usize, usize) {
    let mut points = input
        .lines()
        .enumerate()
        .flat_map(|(y, line)| {
            line.char_indices()
                .filter_map(move |(x, c)| if c == '@' { Some((x, y)) } else { None })
        })
        .collect::<HashSet<_>>();
    let initial = points.len();
    let mut changed = step(&mut points);
    let part1 = initial - points.len();
    while changed {
        changed = step(&mut points);
    }
    (part1, initial - points.len())
}

fn step(points: &mut HashSet<(usize, usize)>) -> bool {
    let old_points = points.clone();
    points
        .extract_if(|(x, y)| {
            (x.saturating_sub(1)..=x.saturating_add(1))
                .flat_map(|x| (y.saturating_sub(1)..=y.saturating_add(1)).map(move |y| (x, y)))
                .filter(|point| old_points.contains(point))
                .count()
                <= 4
        })
        .last()
        .is_some()
}
